diff --git a/buildslave/commands/git.py b/buildslave/commands/git.py
index b6c80e3..3a81126 100644
--- a/buildslave/commands/git.py
+++ b/buildslave/commands/git.py
@@ -89,17 +89,17 @@ class Git(SourceBaseCommand):
         return True
 
     def _cleanSubmodules(self, res):
-        command = ['submodule', 'foreach', 'git', 'clean', '-f', '-d']
+        command = ['submodule', 'foreach', '--recursive', 'git', 'clean', '-f', '-d']
         if self.ignore_ignores:
             command.append('-x')
         return self._dovccmd(command)
 
     def _updateSubmodules(self, res):
-        return self._dovccmd(['submodule', 'update'], self._cleanSubmodules)
+        return self._dovccmd(['submodule', 'update', '--init', '--recursive'], self._cleanSubmodules)
 
     def _initSubmodules(self, res):
         if self.submodules:
-            return self._dovccmd(['submodule', 'init'], self._updateSubmodules)
+            return self._dovccmd(['submodule', 'sync'], self._updateSubmodules)
         else:
             return defer.succeed(0)
 
diff --git a/buildslave/test/unit/test_commands_git.py b/buildslave/test/unit/test_commands_git.py
index 95537a3..c3dfeda 100644
--- a/buildslave/test/unit/test_commands_git.py
+++ b/buildslave/test/unit/test_commands_git.py
@@ -621,15 +621,15 @@ class TestGit(SourceCommandTestMixin, unittest.TestCase):
                    self.basedir_workdir,
                    sendRC=False, timeout=120, usePTY=False)
             + 0,
-            Expect(['path/to/git', 'submodule', 'init'],
+            Expect(['path/to/git', 'submodule', 'sync'],
                    self.basedir_workdir,
                    sendRC=False, timeout=120, usePTY=False)
             + 0,
-            Expect(['path/to/git', 'submodule', 'update'],
+            Expect(['path/to/git', 'submodule', 'update', '--init', '--recursive'],
                    self.basedir_workdir,
                    sendRC=False, timeout=120, usePTY=False)
             + 0,
-            Expect(['path/to/git', 'submodule', 'foreach',
+            Expect(['path/to/git', 'submodule', 'foreach', '--recursive',
                     'git', 'clean', '-f', '-d', '-x'],
                    self.basedir_workdir,
                    sendRC=False, timeout=120, usePTY=False)
